packages/dart/npt_flutter/lib/features/profile_list/bloc/profile_list_bloc.dart

```diff
@@ -97,4 +126,194 @@ class ProfileListBloc extends LoggingBloc<ProfileListEvent, ProfileListState> {
+    final sortedProfiles = await _sortProfiles(
+      currentState.profiles,
+      event.sortColumn,
+      newSortOrder,
+    );
+
+    App.log('Sorted ${sortedProfiles.length} profiles'.loggable);
+    emit(
+      currentState.copyWith(
+        profiles: sortedProfiles,
+        sortColumn: event.sortColumn,
+        sortOrder: newSortOrder,
+      ),
+    );
+  }
+
+  Future<void> _onSearch(
+    ProfileListSearchEvent event,
+    Emitter<ProfileListState> emit,
+  ) async {
+    //TODO: Implement search functionality
+  }
+
+  // Helper Functions below this line
+
+  // Helper function to sort profiles based on column and order
+  Future<List<String>> _sortProfiles(
+    Iterable<String> uuids,
+    SortColumn sortColumn,
+    SortOrder sortOrder,
+  ) async {
+    final profileBlocList = uuids
+        .map((uuid) => _profileCacheCubit.getProfileBloc(uuid))
+        .toList();
+
+    final profileDataList = profileBlocList.map((profileBloc) {
+      final profileState = profileBloc.state;
+
+      if (profileState is ProfileLoadedState) {
+        App.log('Got profile ${profileState.uuid} for sorting'.loggable);
+        return (
+          uuid: profileState.uuid,
+          profile: profileState.profile,
+          state: profileState,
+        );
+      } else {
+        App.log(
+          'Failed to get profile ${profileState.uuid} for sorting: ${profileState.runtimeType}'
+              .loggable,
+        );
+        return (uuid: profileState.uuid, profile: null, state: profileState);
+      }
+    }).toList();
+
+    // Get favorites to pin
+    final context = App.navState.currentContext;
+    final settingsBloc = context?.read<SettingsBloc>();
+    final shouldPinFavorites = settingsBloc?.state is SettingsLoaded
+        ? (settingsBloc!.state as SettingsLoaded).settings.pinFavorites
+        : false;
+
+    App.log('Should pin favorites: $shouldPinFavorites'.loggable);
+
+    if (shouldPinFavorites) {
+      // get favorite uuids
+      final favoriteBloc = context?.read<FavoriteBloc>();
+      final favoriteUuids = <String>{};
+      if (favoriteBloc?.state is FavoritesLoaded) {
+        for (var favorite
+            in (favoriteBloc!.state as FavoritesLoaded).favorites) {
+          favoriteUuids.addAll(favorite.profileIds);
+        }
+      }
+      // Separate and Sort favorites and non-favorites
+      final favorites = profileDataList
+          .where(
+            (data) => data.profile != null && favoriteUuids.contains(data.uuid),
+          )
+          .toList();
+      _sortProfileDataList(favorites, sortColumn, sortOrder);
+      final nonFavorites = profileDataList
+          .where(
+            (data) =>
+                data.profile != null && !favoriteUuids.contains(data.uuid),
+          )
+          .toList();
+      _sortProfileDataList(nonFavorites, sortColumn, sortOrder);
+
+      return [
+        ...favorites.map((e) => e.uuid),
+        ...nonFavorites.map((e) => e.uuid),
+      ];
```

Comment:
sort favourite and non-favourite profiles separately then return a merged list with favourite profiles at the top of the list.